### DISCUSSION TITLE
Conditional Top Layer Patterns

### DISCUSSION THREAD

**Discussion by isaacST08** [OrcaSlicer/OrcaSlicer#4997] [Ideas]
I believe that it would be very beneficial to be able to choose the top layer (and maybe bottom layer?) pattern based on the area of the shape. For example, for thin and long areas a concentric infill pattern results in a more efficient top layer due to the decreased number of direction changes compared to monotonic. But for larger areas the printer can reach a higher speed and the monotonic lines become more efficient where as the concentric pattern starts to generate odd shapes that might be less ideal for speed, appearance, or both. 
A calculation could be made to create a threshold value, something like: Calculate the perimeter of the shape in question. Calculate the maximum area of a square (or circle for complete maximum area but I think a square would be a more appropriate shape benchmark for printing) with the same perimeter. If the area is less than a user defined percentage of that maximum area then use concentric (or whatever pattern the user chooses) and if the area is greater than the chosen threshold then use monotonic (or whatever the user wants).
A minimum constant area threshold could also be set, for example 30 mm^2.

Side note: I feel this logic could also be easily adapted to let users choose where they want ironing on their part. Small areas are typically less than ideal for ironing and waste more time than they bring benefit, but large areas are exactly where I would want to enable ironing.

I would start messing around with a clone of the code but I'm not familiar with the layout of the project and don't know where to start poking so if someone knows where the top layers are generated in the code I would appreciate someone pointing me that way.

I look forward to both growing this idea and hearing any fundamental problems that make this more trouble than its worth.

TLDR: Create an area threshold option for different top layer patterns.

**Comment by isaacST08** [comment]
![13Screenshot_2024-04-13_16:03:37_0001](https://github.com/SoftFever/OrcaSlicer/assets/106057977/93d617fd-6d1f-42c6-ba56-4ea0c5dda776)
![13Screenshot_2024-04-13_16:04:01_0001](https://github.com/SoftFever/OrcaSlicer/assets/106057977/41d17b8a-f55c-49a8-9eb0-3349a3a7e368)
For example here is the same benchy viewed in log layer time. In monotonic (top), the deck of the benchy prints faster, the deck is by far the largest area top surface of the model and benefits from the monotonic top layer pattern. But the facets of the roof and the brim at the back of the benchy benefit more from the concentric top layer pattern (bottom) over the monotonic pattern (top). 
The back brim of the benchy can be aided by turning off "only one wall on top layers" but this has minimal impact on the rest of the areas in question and changing the pattern is far more effective.
